### PR TITLE
New local ai quickstart

### PR DESCRIPTION
**Pull request by alexwolfmsft**
## Summary

Describe your changes here.

Fixes #Issue_Number (if available)

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/ai/quickstarts/quickstart-local-ai.md](https://github.com/dotnet/docs/blob/d228e489cbf0292dbfd2d78641268a3bc337cfea/docs/ai/quickstarts/quickstart-local-ai.md) | [Chat with a local AI model using .NET and Semantic Kernel](https://review.learn.microsoft.com/en-us/dotnet/ai/quickstarts/quickstart-local-ai?branch=pr-en-us-41317) |

### CODE REVIEW

File: docs/ai/quickstarts/quickstart-local-ai.md [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+                        .AddOpenAIChatCompletion(
+                            modelId: "phi3:mini",
+                            endpoint: new Uri("http://localhost:11434"),
+                            apiKey: String.Empty)
+                        .Build();
+    
+    var aiChatService = kernel.GetRequiredService<IChatCompletionService>();
+    var chatHistory = new ChatHistory();
+    
+    while(true)
```
Comment (IEvangelist) [inline docs/ai/quickstarts/quickstart-local-ai.md:90]:
```suggestion
    while (true)
```

File: docs/ai/quickstarts/quickstart-local-ai.md [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    {
+        // Get user prompt and add to chat history
+        Console.WriteLine("Your prompt:");
+        var userPrompt = Console.ReadLine();
+        chatHistory.Add(new ChatMessageContent(AuthorRole.User, userPrompt));
+    
+        // Stream the AI response and add to chat history
+        Console.WriteLine("AI Response:");
+        var response = String.Empty;
```
Comment (IEvangelist) [inline docs/ai/quickstarts/quickstart-local-ai.md:99]:
```suggestion
        var response = "";
```

File: docs/ai/quickstarts/quickstart-local-ai.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    using Microsoft.SemanticKernel;
+    using Microsoft.SemanticKernel.ChatCompletion;
+    
+    // Create a kernel with OpenAI chat completion
+    #pragma warning disable SKEXP0010
+    Kernel kernel = Kernel.CreateBuilder()
+                        .AddOpenAIChatCompletion(
+                            modelId: "phi3:mini",
+                            endpoint: new Uri("http://localhost:11434"),
+                            apiKey: String.Empty)
```
Comment (IEvangelist) [inline docs/ai/quickstarts/quickstart-local-ai.md:84]:
```suggestion
                            apiKey: "")
```

File: docs/ai/quickstarts/quickstart-local-ai.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    {
+        // Get user prompt and add to chat history
+        Console.WriteLine("Your prompt:");
+        var userPrompt = Console.ReadLine();
+        chatHistory.Add(new ChatMessageContent(AuthorRole.User, userPrompt));
+    
+        // Stream the AI response and add to chat history
+        Console.WriteLine("AI Response:");
+        var response = String.Empty;
+        await foreach(var item in aiChatService.GetStreamingChatMessageContentsAsync(chatHistory))
```
Comment (IEvangelist) [inline docs/ai/quickstarts/quickstart-local-ai.md:100]:
```suggestion
        await foreach(var item in 
            aiChatService.GetStreamingChatMessageContentsAsync(chatHistory))
```

File: docs/ai/quickstarts/quickstart-local-ai.md [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+    ```csharp
+    using Microsoft.SemanticKernel;
+    using Microsoft.SemanticKernel.ChatCompletion;
+    
+    // Create a kernel with OpenAI chat completion
+    #pragma warning disable SKEXP0010
+    Kernel kernel = Kernel.CreateBuilder()
+                        .AddOpenAIChatCompletion(
+                            modelId: "phi3:mini",
+                            endpoint: new Uri("http://localhost:11434"),
```
Comment (IEvangelist) [inline docs/ai/quickstarts/quickstart-local-ai.md:83]: Why use this endpoint, it's not mentioned anywhere? Does `ollama` listening on this port or something?

Comment (alexwolfmsft) [reply docs/ai/quickstarts/quickstart-local-ai.md:83]: It's the default endpoint that Ollama starts up on after a standard install - I will add a note about this, good catch.

File: docs/ai/quickstarts/quickstart-local-ai.md [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    ```
+
+    The preceding code accomplishes the following tasks:
+    - Creates a `Kernel` object and uses it to retrieve a chat completion service.
+    - Creates a `ChatHistory` object to store the messages between the user and the AI model.
+    - Retrieves a prompt from the user and stores it in the `ChatHistory`.
+    - Sends the chat data to the AI model to generate a response.
+
+    > [!NOTE]
+    > Ollama runs on portal 11434 by default, which is why the AI model endpoint is set to `http://localhost:11434`.
```
Comment (IEvangelist) [inline docs/ai/quickstarts/quickstart-local-ai.md:117]:
```suggestion
    > Ollama runs on port 11434 by default, which is why the AI model endpoint is set to `http://localhost:11434`.
```

Comment (alexwolfmsft) [reply docs/ai/quickstarts/quickstart-local-ai.md:117]: You beat me to it

Comment (alexwolfmsft) [reply docs/ai/quickstarts/quickstart-local-ai.md:117]: You may now have cake.

**Review by IEvangelist — APPROVED** [review]

**Comment by alexwolfmsft** [comment]
> But when we tried it, it gives following error in await foreach (var item in aiChatService.GetStreamingChatMessageContentsAsync(chatHistory))
> 
> **_System.ClientModel.ClientResultException: 'Service request failed. Status: 404 (Not Found)_**
> 
> On the same system, if we try the sample from OlllamaSharp - that works fine on the same model + same port.
> 
> If we browse to the http://localhost:11434 it shows "Ollama is running"
> 
> Any suggestions?

Thank you for surfacing this - you can find an updated code example here:
https://github.com/dotnet/docs/blob/main/docs/ai/quickstarts/snippets/local-ai/Program.cs

This approach uses the Microsoft.SemanticKernel.Connectors.Ollama package for a better connection experience to Ollama, so you'll need to add that to your project as well.

**Pull request merged**